### ISSUE TITLE
User Error/ Question: VSCode Terminal keeps Popping up when executing a cell

### ISSUE THREAD

**Issue by jlewi** [stateful/runme#628]
Here's a video illustrating that when I execute a cell. The terminal keeps popping up.

https://github.com/stateful/runme/assets/777219/4ac12c13-d0ec-4b2f-99cc-07622df89c3e

@sourishkrout do you know if this is a vscode setting? A runme setting?

**Comment by jlewi** [comment]
I think I partially figured it out. It looks like the cell is configured to run in interactive mode and that causes the terminal output to pop up. When I uncheck "Run Interactive" it no longer happens.

Is there away to make non-interactive the default?

Interestingly this doesn't seem to happen in the demo videos in the docs
https://docs.runme.dev/how-runme-works/vscode#interactive-mode

**Comment by sourishkrout** [comment]
Definitely odd, @jlewi. The video shows you're using a workspace, is that right? In the sense that you have multiple folders add to a VS Code window? Mainly asking to help my with reproducing the issue.

**Comment by jlewi** [comment]
@sourishkrout Yes. It was in a workspace. But I tried in a non workspace and I had a similar problem. In that case the terminal would pop up and then disappear.

**Comment by jlewi** [comment]
I tried reinstalling the extension and that didn't help. 
I have a 2nd work laptop and on that one its not happening. I also notice that the "preview gist" button is missing.

**Comment by sourishkrout** [comment]
Could you check the markdown document and see if `closeTerminalOnSuccess` isn't set to `false` (as per https://docs.runme.dev/Reference/configuration)?

I'd also suggest checking `@ext:stateful.runme @modified` to see if `runme.terminal.closeOnSuccess` is set to default which is `true`.

Btw, reinstalling extensions does not wipe settings, they keep dangling and reactivate on install.

**Comment by sourishkrout** [comment]
> `closeOnSuccess` is set to true.
> I don't see a `closeTerminalOnSuccess` option for the cell

We removed it from the UI since it's only useful in infrequent corner cases. You can however, add/remove it into the JSON-encoded attributes per cell inside the raw markdown. If you don't see in the raw markdown anywhere it's safe to assume, it's defaulting to `true`.

@jlewi would you be able to generate a GitHub Gist of that markdown file for my to try reproducing it locally?

**Comment by jlewi** [comment]
@sourishkrout yup; I'll do that today.

**Comment by jlewi** [comment]
Here's a gist.
https://gist.github.com/jlewi/60a490edc6b5e18c52257b2f735eb7af

It seems to happen in notebook. I just created a new one.

**Comment by sourishkrout** [comment]
> Here's a gist. https://gist.github.com/jlewi/60a490edc6b5e18c52257b2f735eb7af
> 
> It seems to happen in notebook. I just created a new one.

Can you please check that `runtime.experiments.grpcRunner` is set to default/unmodified? It should be true. Also, if you could change the cell to `echo $RUNME_RUNNER` if the `GrpcRunner` is active, it should output its version.

Based on the "plain notebook" and the fact that it doesn't render a terminal makes me think that somehow the `GrpcRunner` is not being used. The setting above would be the obvious culprit.

However, we also hard-disable the `GrpcRunner` for Windows (non-WSL2) because PowerShell support is incomplete:

https://github.com/stateful/vscode-runme/blob/main/src/extension/kernel.ts#L750-L752

**Comment by jlewi** [comment]
Its using gRPC Runner here's the gist showing its v1
https://gist.github.com/jlewi/18339a0f63f07e43c5e63e5e7ff242a6
I'm on MacOS.

**Comment by jlewi** [comment]
@sourishkrout IUC when a cell is run in interactive mode its supposed to open up a vscode terminal inside the output cell of the notebook in order to make it interactive. It seems like in my case, the terminal window is being opened up in the terminal panel of vscode. Any idea why that might be happening? How does RunMe open a terminal in the notebook output window?

**Comment by sourishkrout** [comment]
> @sourishkrout IUC when a cell is run in interactive mode its supposed to open up a vscode terminal inside the output cell of the notebook in order to make it interactive. It seems like in my case, the terminal window is being opened up in the terminal panel of vscode. Any idea why that might be happening? How does RunMe open a terminal in the notebook output window?

This behavior matches what the Runme MVP (pre grpc runner) used to do before we did terminal multiplexing. However, we already ruled out that in this instance it's running in pre-grpc-runner mode. The terminal is being added to the notebook as distinct cell output.

<img width="931" alt="image" src="https://github.com/user-attachments/assets/1747275b-980a-45e8-8e9d-63cad6d95d42">

> 
> I also notice `Terminal will be reused by tasks, press any key to close it. `. So does that indicate the terminal is not being closed and therefore being reused?

The terminal as in its tty/pty is being reused for multiplexing it's input/output. What that means is one can observe the terminal both inside the notebook and the bottom panel.

What stands out to me is that in all the Gists you've generated we don't get a `exited with 0` which makes me think perhaps the exit code is inconclusive and the fallback behavior might be to open the terminal. Why this happens... unclear. Can you copy and paste all the logs you see in `Output->Runme` please? We could also hop on a hangout/zoom to do some real-time troubleshooting.

See this example including the exit code in contrast to the Gists you've posted: https://gist.github.com/sourishkrout/d4d1fe028f3c1a365d410e0448bfaa93#file-01j24rc9r3k9hcqcjgs1bs6kkv-md

**Comment by jlewi** [comment]
Output->RunMe

```
[2024-07-16T12:45:48.277Z] INFO Runme: Activating Extension
[2024-07-16T12:45:48.281Z] INFO Runme(RunmeTaskProvider): Walking 3 directories/repos...
[2024-07-16T12:45:48.312Z] INFO Runme(KernelServer): AI logs enabled
[2024-07-16T12:45:48.326Z] INFO Runme(KernelServer): Server process #15937 started at localhost:7863
[2024-07-16T12:45:48.364Z] INFO Runme(Kernel): Requesting new runner environment.
[2024-07-16T12:45:48.374Z] ERROR Runme(Kernel): Error monitoring env store Request message serialization failure: 
[2024-07-16T12:45:48.376Z] INFO Runme(Kernel): New runner environment assigned with session ID: 01J2XS7QEFPQDDVEHVPBFY95KC
[2024-07-16T12:45:48.381Z] INFO Runme: Extension successfully activated
[2024-07-16T12:45:48.713Z] INFO Runme(RunmeTaskProvider): Finished walk.
[2024-07-16T12:46:58.122Z] ERROR Runme(Kernel): Unknown kernel event type: common:getState
[2024-07-16T12:47:41.530Z] ERROR Runme(Kernel): Unknown kernel event type: common:getState
[2024-07-16T12:47:58.046Z] ERROR Runme(Kernel): Unknown kernel event type: common:getState
[2024-07-16T12:47:58.058Z] ERROR Runme(Kernel): Unknown kernel event type: common:getState
[2024-07-16T12:47:58.288Z] ERROR Runme(Kernel): Unknown kernel event type: common:getState
[2024-07-16T12:47:58.613Z] ERROR Runme(Kernel): Unknown kernel event type: common:getState
[2024-07-16T12:47:58.846Z] ERROR Runme(Kernel): Unknown kernel event type: common:getState
[2024-07-16T12:47:58.936Z] ERROR Runme(Kernel): Unknown kernel event type: common:getState
[2024-07-16T12:47:59.016Z] ERROR Runme(Kernel): Unknown kernel event type: common:getState
[2024-07-16T12:47:59.023Z] ERROR Runme(Kernel): Unknown kernel event type: common:getState
[2024-07-16T12:47:59.172Z] ERROR Runme(Kernel): Unknown kernel event type: common:getState
[2024-07-16T12:47:59.187Z] ERROR Runme(Kernel): Unknown kernel event type: common:getState
[2024-07-16T12:47:59.205Z] ERROR Runme(Kernel): Unknown kernel event type: common:getState
```

I noticed in Output -> Tasks the following

```
Error: there is no registered task type 'customExecution'. Did you miss installing an extension that provides a corresponding task provider?
```

**Comment by jlewi** [comment]
@sourishkrout got to the bottom of this.
It was user error.
I had unchecked "Runme Terminal: Interactive".  
To fix it I needed to
1. Enable it by checking the box (restoring its default)
2. Run the Reload Window command so it took effect

I think I skipped step 2 in earlier experiments which is why it didn't work.
